From 4558cb476ed9030e82838b733eb69d63052e66cf Mon Sep 17 00:00:00 2001
From: Felix Fietkau <nbd@nbd.name>
Date: Thu, 16 Mar 2023 19:31:13 +0530
Subject: [PATCH 104/281] OpenWrt: 259-regmap_dynamic.patch

kernel: move regmap bloat out of the kernel image if it is only being used in modules

lede-commit: 96f39119815028073583e4fca3a9c5fe9141e998
Signed-off-by: Felix Fietkau <nbd@nbd.name>

Change-Id: I284a66d1743616506500c246648fc88cd79ffe2a
Signed-off-by: Ram Chandra Jangir <quic_rjangir@quicinc.com>
---
 drivers/base/regmap/Kconfig  | 17 +++++++++++++++--
 drivers/base/regmap/Makefile | 12 ++++++++----
 drivers/base/regmap/regmap.c |  3 +++
 include/linux/regmap.h       |  2 +-
 4 files changed, 27 insertions(+), 7 deletions(-)

diff --git a/drivers/base/regmap/Kconfig b/drivers/base/regmap/Kconfig
index 159bac6c5046..3b7f5b1d34fb 100644
--- a/drivers/base/regmap/Kconfig
+++ b/drivers/base/regmap/Kconfig
@@ -4,10 +4,9 @@
 # subsystems should select the appropriate symbols.
 
 config REGMAP
-	default y if (REGMAP_I2C || REGMAP_SPI || REGMAP_SPMI || REGMAP_W1 || REGMAP_AC97 || REGMAP_MMIO || REGMAP_IRQ || REGMAP_SOUNDWIRE || REGMAP_SOUNDWIRE_MBQ || REGMAP_SCCB || REGMAP_I3C || REGMAP_SPI_AVMM || REGMAP_MDIO)
 	select IRQ_DOMAIN if REGMAP_IRQ
 	select MDIO_BUS if REGMAP_MDIO
-	bool
+	tristate
 
 config REGCACHE_COMPRESSED
 	select LZO_COMPRESS
@@ -15,53 +14,67 @@ config REGCACHE_COMPRESSED
 	bool
 
 config REGMAP_AC97
+	select REGMAP
 	tristate
 
 config REGMAP_I2C
+	select REGMAP
 	tristate
 	depends on I2C
 
 config REGMAP_SLIMBUS
+	select REGMAP
 	tristate
 	depends on SLIMBUS
 
 config REGMAP_SPI
+	select REGMAP
 	tristate
 	depends on SPI
 
 config REGMAP_SPMI
+	select REGMAP
 	tristate
 	depends on SPMI
 
 config REGMAP_W1
+	select REGMAP
 	tristate
 	depends on W1
 
 config REGMAP_MDIO
+	select REGMAP
 	tristate
 
 config REGMAP_MMIO
+	select REGMAP
 	tristate
 
 config REGMAP_IRQ
+	select REGMAP
 	bool
 
 config REGMAP_SOUNDWIRE
+	select REGMAP
 	tristate
 	depends on SOUNDWIRE
 
 config REGMAP_SOUNDWIRE_MBQ
+	select REGMAP
 	tristate
 	depends on SOUNDWIRE
 
 config REGMAP_SCCB
+	select REGMAP
 	tristate
 	depends on I2C
 
 config REGMAP_I3C
+	select REGMAP
 	tristate
 	depends on I3C
 
 config REGMAP_SPI_AVMM
+	select REGMAP
 	tristate
 	depends on SPI
diff --git a/drivers/base/regmap/Makefile b/drivers/base/regmap/Makefile
index 11facb32a027..092ccf7e95c9 100644
--- a/drivers/base/regmap/Makefile
+++ b/drivers/base/regmap/Makefile
@@ -2,10 +2,14 @@
 # For include/trace/define_trace.h to include trace.h
 CFLAGS_regmap.o := -I$(src)
 
-obj-$(CONFIG_REGMAP) += regmap.o regcache.o
-obj-$(CONFIG_REGMAP) += regcache-rbtree.o regcache-flat.o
-obj-$(CONFIG_REGCACHE_COMPRESSED) += regcache-lzo.o
-obj-$(CONFIG_DEBUG_FS) += regmap-debugfs.o
+regmap-core-objs = regmap.o regcache.o regcache-rbtree.o regcache-flat.o
+ifdef CONFIG_DEBUG_FS
+regmap-core-objs += regmap-debugfs.o
+endif
+ifdef CONFIG_REGCACHE_COMPRESSED
+regmap-core-objs += regcache-lzo.o
+endif
+obj-$(CONFIG_REGMAP) += regmap-core.o
 obj-$(CONFIG_REGMAP_AC97) += regmap-ac97.o
 obj-$(CONFIG_REGMAP_I2C) += regmap-i2c.o
 obj-$(CONFIG_REGMAP_SLIMBUS) += regmap-slimbus.o
diff --git a/drivers/base/regmap/regmap.c b/drivers/base/regmap/regmap.c
index 7de1f27d0323..788aa6110153 100644
--- a/drivers/base/regmap/regmap.c
+++ b/drivers/base/regmap/regmap.c
@@ -9,6 +9,7 @@
 #include <linux/device.h>
 #include <linux/slab.h>
 #include <linux/export.h>
+#include <linux/module.h>
 #include <linux/mutex.h>
 #include <linux/err.h>
 #include <linux/property.h>
@@ -3511,3 +3512,5 @@ static int __init regmap_initcall(void)
 	return 0;
 }
 postcore_initcall(regmap_initcall);
+
+MODULE_LICENSE("GPL");
diff --git a/include/linux/regmap.h b/include/linux/regmap.h
index ca3434dca3a0..6a30121a6b5d 100644
--- a/include/linux/regmap.h
+++ b/include/linux/regmap.h
@@ -180,7 +180,7 @@ struct reg_sequence {
 	__ret ?: __tmp; \
 })
 
-#ifdef CONFIG_REGMAP
+#if IS_REACHABLE(CONFIG_REGMAP)
 
 enum regmap_endian {
 	/* Unspecified -> 0 -> Backwards compatible default */
-- 
2.17.1

